cranelift/codegen/src/verifier/mod.rs

```diff
@@ -2040,12 +2042,64 @@ impl<'a> Verifier<'a> {
         Ok(())
     }
 
+    fn verify_signature(
+        &self,
+        loc: impl Into<AnyEntity>,
+        data: &Signature,
+        errors: &mut VerifierErrors,
+    ) -> VerifierStepResult {
+        if data.call_conv == CallConv::Patchable {
+            if data.params.len() > 4 {
+                return errors.fatal((
+                    loc,
+                    "signature with patchable ABI does not allow more than four arguments"
+                        .to_string(),
+                ));
+            }
+            if data.returns.len() > 0 {
+                return errors.fatal((
+                    loc,
+                    "signature with patchable ABI does not allow any returns".to_string(),
+                ));
+            }
+            for param in &data.params {
+                if param.value_type != crate::ir::types::I32
+                    && param.value_type != crate::ir::types::I64
+                {
+                    return errors.fatal((
+                        loc,
+                        "signature with patchable ABI does not allow non-I32/I64 arguments"
+                            .to_string(),
+                    ));
+                }
+                if param.extension != ArgumentExtension::None {
+                    return errors.fatal((
+                        loc,
+                        "signature with patchable ABI does not allow sign/zero-extended arguments"
+                            .to_string(),
+                    ));
+                }
+            }
+        }
+        Ok(())
+    }
```

Comment:
Okay great, this is exactly what I was hoping would be here. We should document these constraints in the `ir::CallConv` doc comments for cranelift embedders tho.